### ISSUE TITLE
Compilation on Windows?

### ISSUE THREAD

**Issue by Dev-iL** [exoclime/HELIOS-K#1]
**TL;DR - Wanted to compile this on Win10 using VS, managed to get it partially working, requesting an official (_supported_) way to do it.**

------------
Unfortunately for myself, I'm not very familiar with this sort of tasks, so I had to fix compiler errors incrementally, probably crudely, though with some success (got `heliosk.cu` to compile). I'm providing the steps I took in case anybody else needs this:

1) Created new header files in the `<HELIOS-K ROOT>/sys` folder:
- `time.h`:
```c
#pragma once
 
#ifndef _TIMES_H
 
#include "sys/times.h"
 
#endif
```
- `times.h`:
```c
#ifndef _TIMES_H
#define _TIMES_H

#ifdef _WIN32
#include <sys/timeb.h>
#include <sys/types.h>
#include <winsock2.h>

int gettimeofday(struct timeval* t,void* timezone)
{       struct _timeb timebuffer;
        _ftime( &timebuffer );
        t->tv_sec=timebuffer.time;
        t->tv_usec=1000*timebuffer.millitm;
		return 0;
}

// from linux's sys/times.h

//#include <features.h>

#define __need_clock_t
#include <time.h>


/* Structure describing CPU time used by a process and its children.  */
struct tms {
    clock_t tms_utime;          /* User CPU time.  */
    clock_t tms_stime;          /* System CPU time.  */

    clock_t tms_cutime;         /* User CPU time of dead children.  */
    clock_t tms_cstime;         /* System CPU time of dead children.  */
};

struct timezone {
	int tz_minuteswest; /* minutes west of Greenwich */
	int tz_dsttime; /* type of DST correction */
};

/* Store the CPU time used by this process and all its
   dead children (and their dead children) in BUFFER.
   Return the elapsed real time, or (clock_t) -1 for errors.
   All times are in CLK_TCKths of a second.  */
clock_t times (struct tms *__buffer) {

	__buffer->tms_utime = clock();
	__buffer->tms_stime = 0;
	__buffer->tms_cstime = 0;
	__buffer->tms_cutime = 0;
	return __buffer->tms_utime;
}

typedef long long suseconds_t ;

#endif
#endif
```

 2. In `heliosk.cu`, replaced the lines 
```c
#include <sys/time.h>
#include <sys/resource.h>
```
   with
```c
#ifdef __GNUC__
  #define LINUX
#else
  #define WINDOWS
#endif

#ifdef WINDOWS
  #include "sys/time.h"
#else
  #include <sys/time.h>
  #include <sys/resource.h>
#endif 
```

 3. Used this as a `Makefile`:
```make
SM=35
ARCH = -arch sm_$(SM)
source = heliosk.cu
headers = define.h ISO.h resample.h host.h voigt.h
all: heliosk
heliosk: $(source) $(headers)
	nvcc $(ARCH) -o heliosk $(source)
#	nvcc --ptxas-options=-v $(ARCH) -o heliosk $(source)
#	nvcc --compiler-options -Wall $(ARCH) -o heliosk $(source)
```

 4. Compiled it using:
```cmd
nmake SM=30
```

**Comment by sigrimm** [comment]
Thank you for this request.

I have changed the timer metrics in heliok to use CUDA events instead of gettimeofday. This does not need to include the sys/time.h file anymore and can know be compiled also in Windows.

**Comment by Dev-iL** [comment]
Thank you for your modifications so far, I am indeed able to compile the `heliosk` object without any manual modifications. However, VS complains about the following line:

https://github.com/exoclime/HELIOS-K/blob/c258835c21d3c71b696e1cc0726db877c4300d49/prepare.cpp#L261

which gives a compilation error since apparently one cannot allocate memory based on a runtime value, w/o using `malloc`. Replacing it with the following appears to remove the error:
```cpp
int* nLines = (int*)malloc(m.nFiles * sizeof(int));
```
Also, perhaps it's just an issue of the toolchain I'm using, but `nmake` doesn't like the syntax of the Makefile (specifically, `:=`). If this is some Linux-specifix syntax, would it be possible to surround it with some OS-checking code, as shown e.g. [here](https://stackoverflow.com/q/714100/)?

Finally, there are some strange indentation anomalies in the source and header files. (I'd be happy to submit a PR and fix this if needed.)

**Comment by sigrimm** [comment]
Thanks again!

I have changed the allocation in prepare.cpp.

I have included an OS-checking in the Makefile. However this does not work in a Windows Command prompt with nmake. For this reason, I have added a sepparate Makefile for Windows, called MakefileW.
compilation with "nmake -f MakefileW SM=50" works on my machine.
Or by using Cygwin, make can directly be used.

I cleaned up some indentations. If you find more anomalies, I would be happy to correct them as well.

**Comment by Dev-iL** [comment]
Hi, 

I suppose my system was setup somewhat differently from yours, so running `cl prepare.cpp` (the command which appears in `MakefileW`) failed for me with the error:

```none
.\prepare.cpp(1): fatal error C1034: stdlib.h: no include path set
```

because apparently the compiler doesn't know about any include paths. To fix this, one must run the command `vcvars32` to setup the compilation environment. In other words, I suggest making the following modification to the makefile:

```diff
- cl prepare.cpp
+ vcvars32 & cl prepare.cpp
```

(ditto for `prepareExomol`).

**Comment by sigrimm** [comment]
Hi,

Thanks for the suggestion, but I would not include that in the Makefile, because the name of this compiler file depends on the system setup. Sometimes it can be called vcvars64.bat, vsvars32.bat or similar.

I will leave it up to the user to load the correct compiler path. In most cases, it can be achieved by a simple call in the command promt, similar to:
call "C:\Program Files (x86)\Microsoft Visual Studio 14.0\Common7\Tools\vsvars32.bat",
where the exact path and file name must eventually be changed. 

I hope this solution works for you.

**Comment by Dev-iL** [comment]
Hi,

Either way is fine. What's important to me is that both suggestions are kept in the repository for posterity. I'm closing the issue now.

Would you add a note to the readme that if compilation fails with aforementioned error, "please see issue #1"?

**Comment by sigrimm** [comment]
Done. Thank you
